API/Auth/DlcsBasicAuthenticationHandler.cs

```diff
@@ -0,0 +1,147 @@
+            int? resourceCustomerId = null;
+            if (Request.RouteValues.TryGetValue("customerId", out var customerIdRouteVal))
+            {
+                if (int.TryParse(customerIdRouteVal.ToString(), out int result))
+                {
+                    resourceCustomerId = result;
+                }
+            }
+            var authHeader = Encoding.UTF8.GetString(Convert.FromBase64String(headerValue.Parameter));
+            string[] keyAndSecret = authHeader.Split(':');
+            var apiCaller = new {Key = keyAndSecret[0], Secret = keyAndSecret[1]};
+            var customerForKey = await customerRepository.GetCustomerForKey(apiCaller.Key, resourceCustomerId);
+            if (customerForKey == null)
+            {
+                return AuthenticateResult.Fail("No customer found for this key that is permitted to access this resource");
+            }
+
+            if (apiCaller.Secret != deliveratorApiAuth.GetApiSecret(customerForKey, Options.Salt, apiCaller.Key))
+            {
+                return AuthenticateResult.Fail("Invalid credentials");
+            }
+
+            // We still have some checks to do before authenticating this user.
+            // Some of these could be classed as authorisation rather than authentication, though they are not user-specific.
+            if (resourceCustomerId.HasValue)
+            {
+                // the request is for a particular customer's resource (e.g., an asset)
+                if (customerForKey.Id != resourceCustomerId.Value)
+                {
+                    // ... but the requester is not this customer. Only proceed if they are admin.
+                    if (!customerForKey.Administrator)
+                    {
+                        return AuthenticateResult.Fail("Only admin user may access this customer's resource.");
+                    }
+                    Logger.LogInformation("Admin key accessing a customer's resource");
+                }
+            }
+            else
+            {
+                // The request isn't for a customer resource (i.e., under the path /customers/n/).
+                // Only proceed if they are admin.
+                if (!customerForKey.Administrator)
+                {
+                    return AuthenticateResult.Fail("Only admin user may access this shared resource");
+                }
+                Logger.LogInformation("Admin key accessing a shared resource");
+            }
+            
+            // At this point our *authentication* has passed.
+            // Downstream handlers may still refuse to *authorise* the request for other reasons
+            // (it can still end up a 401)
+            Logger.LogInformation("Authentication passed");
+
+            var claims = new List<Claim>
+            {
+                new (ClaimTypes.Name, customerForKey.Name), // TODO - should these be the other way round?
+                new (ClaimTypes.NameIdentifier, apiCaller.Key), // ^^^

```

Comment:
I think `new (ClaimTypes.Name, customerForKey.Name)` is correct, iirc `ClaimTypes.Name` is for display. `ClaimTypes.NameIdentifier` is for immutable identifier. I wonder if apikey is neither of those..